src/ggml-rpc.cpp

```diff
@@ -0,0 +1,622 @@
+    serialize_tensor(src, protobuf_src);
+    ggml::Tensor * protobuf_dst = request.mutable_dst();
+    serialize_tensor(dst, protobuf_dst);
+    ggml::CopyTensorReply reply;
+    grpc::ClientContext context;
+    grpc::Status status = ctx->stub->CopyTensor(&context, request, &reply);
+    GGML_ASSERT(status.ok());
+    return reply.result();
+}
+
+GGML_CALL static void ggml_backend_rpc_buffer_clear(ggml_backend_buffer_t buffer, uint8_t value) {
+    GGML_PRINT_DEBUG("buffer clear\n");
+    ggml_backend_rpc_buffer_context * ctx = (ggml_backend_rpc_buffer_context *)buffer->context;
+    ggml::BufferClearRequest request;
+    request.set_bufptr(ctx->remote_ptr);
+    request.set_value(value);
+    ggml::BufferClearReply reply;
+    grpc::ClientContext context;
+    grpc::Status status = ctx->stub->BufferClear(&context, request, &reply);
+    GGML_ASSERT(status.ok());
+}
+
+static ggml_backend_buffer_i ggml_backend_rpc_buffer_interface = {
+    /* .get_name        = */ ggml_backend_rpc_buffer_get_name,
+    /* .free_buffer     = */ ggml_backend_rpc_buffer_free_buffer,
+    /* .get_base        = */ ggml_backend_rpc_buffer_get_base,
+    /* .init_tensor     = */ ggml_backend_rpc_buffer_init_tensor,
+    /* .set_tensor      = */ ggml_backend_rpc_buffer_set_tensor,
+    /* .get_tensor      = */ ggml_backend_rpc_buffer_get_tensor,
+    /* .cpy_tensor      = */ ggml_backend_rpc_buffer_cpy_tensor,
+    /* .clear           = */ ggml_backend_rpc_buffer_clear,
+    /* .reset           = */ NULL,
+};
+
+// rpc buffer type
+struct ggml_backend_rpc_buffer_type_context {
+    std::shared_ptr<ggml::Backend::Stub> stub;
+    std::string name;
+};
+
+GGML_CALL static const char * ggml_backend_rpc_buffer_type_name(ggml_backend_buffer_type_t buft) {
+    GGML_PRINT_DEBUG("buffer type name\n");
+    ggml_backend_rpc_buffer_type_context * buft_ctx = (ggml_backend_rpc_buffer_type_context *)buft->context;
+    return buft_ctx->name.c_str();
+}
+
+GGML_CALL static ggml_backend_buffer_t ggml_backend_rpc_buffer_type_alloc_buffer(ggml_backend_buffer_type_t buft, size_t size) {
+    GGML_PRINT_DEBUG("alloc buffer, size: %zu\n", size);
+    ggml_backend_rpc_buffer_type_context * buft_ctx = (ggml_backend_rpc_buffer_type_context *)buft->context;
+    ggml::AllocateBufferRequest request;
+    request.set_size(size);
+    ggml::AllocateBufferReply reply;
+    grpc::ClientContext context;
+    grpc::Status status = buft_ctx->stub->AllocateBuffer(&context, request, &reply);
+    GGML_ASSERT(status.ok());
+    ggml_backend_buffer_t buffer = (ggml_backend_buffer_t) malloc(sizeof(struct ggml_backend_buffer));
+
+    (*buffer) = (struct ggml_backend_buffer) {
+        /* .interface = */ ggml_backend_rpc_buffer_interface,
+        /* .buft      = */ buft,
+        /* .context   = */ new ggml_backend_rpc_buffer_context{buft_ctx->stub, reply.bufptr(), "RPC Buffer"},
+        /* .size      = */ reply.size(),
+        /* .usage     = */ GGML_BACKEND_BUFFER_USAGE_ANY
+    };
```

Comment:
Even though this should work in most cases since you are using `malloc`, freeing the `ggml_backend_buffer` is the responsibility of `ggml_backend_buffer_free`, and calling `ggml_backend_buffer_init` ensures that the matching allocation function is used.

```suggestion
    ggml_backend_buffer_t buffer = ggml_backend_buffer_init(buft, ggml_backend_rpc_buffer_interface, new ggml_backend_rpc_buffer_context{buft_ctx->stub, reply.bufptr(), "RPC Buffer"}, reply.size());
```

Reply:
Fixed